### ISSUE TITLE
Possible resource leakage on BuiltingFunctions (Capture)

### ISSUE THREAD

**Issue by Altaflux** [schibsted/jslt#11]
I see on the class BuiltingFunctions on the subclass Capture that a static cache exists for JstlPatterns, I suppose this is there to try to optimize capture.

Where I work we are investigating to possibility of using this nice library for our json transformations but on our scenarios the jslts that we use may be on very constant change as they could be injected to the application from an external source we don't control. But while looking at the code one of the concerns we have is that this cache cannot be cleaned and will be kept until a restart.

In my opinion this cache should be somehow bound to the ParseContext and not shared as static so it doesn't leak.

**Comment by larsga** [comment]
You're absolutely right that there is a possible resource leak here. For our uses it has not been a problem, but I agree that it could theoretically be an issue in some cases. I'll look into this and see if I can solve it.

**Comment by larsga** [comment]
Some thinking out loud about how to solve this.

How this could go wrong:

1. The JSLT expressions are generated dynamically from some external source, so that there is a very large number of regular expressions over time, filling up the cache.
2. The regular expression comes from the JSON input, which can lead to the same issue.

Some ways we could implement this:
1. The current global cache. Very efficient, but vulnerable to the two issues above.
2. Have a separate cache per JSLT expression. Vulnerable to issue 2 above.
3. Optimize regular expression function expressions with static regexp into a closure that holds the compiled regexp. Solves issue 1 above, but in case 2 we lose the caching completely.
4. Set a ceiling on the size of the cache, and start evicting entries when it gets too big.

**Comment by larsga** [comment]
I wrote a test case for this that generates 10 million random regexps and runs them through `capture`. The original implemention very quickly balloons up to 7-8 gigabytes of RAM and threatens to make my laptop unresponsive. The new implementation finishes the entire test suite in 12 seconds. So I'm pretty confident we can declare this issue closed.

Let me know if you have more concerns. Also, I'll see if I can put out an 0.1.2 release, since we've accumulated a few bug fixes now.
